translator/trace/protospan_translation.go

```diff
@@ -102,25 +136,171 @@ func AttributeValueToString(attr pdata.AttributeValue, jsonLike bool) string {
-		// OpenCensus attributes cannot represent maps natively. Convert the
-		// map to a JSON-like string.
-		var sb strings.Builder
-		sb.WriteString("{")
-		m := attr.MapVal()
-		first := true
-		m.ForEach(func(k string, v pdata.AttributeValue) {
-			if !first {
-				sb.WriteString(",")
-			}
-			first = false
-			sb.WriteString(fmt.Sprintf("%q:%s", k, AttributeValueToString(v, true)))
-		})
-		sb.WriteString("}")
-		return sb.String()
+		jsonStr, _ := json.Marshal(AttributeMapToMap(attr.MapVal()))
+		return string(jsonStr)
+
+	case pdata.AttributeValueARRAY:
+		jsonStr, _ := json.Marshal(AttributeArrayToSlice(attr.ArrayVal()))
+		return string(jsonStr)
 
 	default:
 		return fmt.Sprintf("<Unknown OpenTelemetry attribute value type %q>", attr.Type())
 	}
+}
+
+// AttributeMapToMap converts an OTLP AttributeMap to a standard go map
+func AttributeMapToMap(attrMap pdata.AttributeMap) map[string]interface{} {
+	rawMap := make(map[string]interface{})
+	attrMap.ForEach(func(k string, v pdata.AttributeValue) {
+		switch v.Type() {
+		case pdata.AttributeValueSTRING:
+			rawMap[k] = v.StringVal()
+		case pdata.AttributeValueINT:
+			rawMap[k] = v.IntVal()
+		case pdata.AttributeValueDOUBLE:
+			rawMap[k] = v.DoubleVal()
+		case pdata.AttributeValueBOOL:
+			rawMap[k] = v.BoolVal()
+		case pdata.AttributeValueNULL:
+			rawMap[k] = nil
+		case pdata.AttributeValueMAP:
+			rawMap[k] = AttributeMapToMap(v.MapVal())
+		case pdata.AttributeValueARRAY:
+			rawMap[k] = AttributeArrayToSlice(v.ArrayVal())
+		}
+	})
+	return rawMap
+}
+
+func AttributeArrayToSlice(attrArray pdata.AnyValueArray) []interface{} {
+	rawSlice := make([]interface{}, 0, attrArray.Len())
+	for i := 0; i < attrArray.Len(); i++ {
+		v := attrArray.At(i)
+		switch v.Type() {
+		case pdata.AttributeValueSTRING:
+			rawSlice = append(rawSlice, v.StringVal())
+		case pdata.AttributeValueINT:
+			rawSlice = append(rawSlice, v.IntVal())
+		case pdata.AttributeValueDOUBLE:
+			rawSlice = append(rawSlice, v.DoubleVal())
+		case pdata.AttributeValueBOOL:
+			rawSlice = append(rawSlice, v.BoolVal())
+		case pdata.AttributeValueNULL:
+			rawSlice = append(rawSlice, nil)
+		default:
+			rawSlice = append(rawSlice, "<Invalid array value>")
+		}
+	}
+	return rawSlice
+}
 
-	// TODO: Add support for ARRAY type.
+// UpsertStringToAttributeMap upserts a string value to the specified key as it's native OTLP type
+func UpsertStringToAttributeMap(key string, val string, dest pdata.AttributeMap, omitSimpleTypes bool) {
+	switch DetermineValueType(val, omitSimpleTypes) {
+	case pdata.AttributeValueINT:
+		iVal, _ := strconv.ParseInt(val, 10, 64)
+		dest.UpsertInt(key, iVal)
+	case pdata.AttributeValueDOUBLE:
+		fVal, _ := strconv.ParseFloat(val, 64)
+		dest.UpsertDouble(key, fVal)
+	case pdata.AttributeValueBOOL:
+		bVal, _ := strconv.ParseBool(val)
+		dest.UpsertBool(key, bVal)
+	case pdata.AttributeValueMAP:
+		var attrs map[string]interface{}
+		err := json.Unmarshal([]byte(val), &attrs)
```

Comment:
I don't think we should try to treat every string that looks like a JSON as a JSON and convert it to a map. This is can be an unexpected behavior. The recommendation in the spec is for emitters to use JSON if they have no other way to represent structured data. But there is no recommendation in the spec for readers to try to treat it as a JSON. I do not think this is Collector's business. If backends want to do this they are free to but we are overloading Collector by doing this here.